cli/profile.go

```diff
@@ -0,0 +1,143 @@
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/urfave/cli/v2"
+)
+
+const profileEnvVar = "VIAM_CLI_PROFILE_NAME"
+
+type addOrUpdateProfileArgs struct {
+	ProfileName string
+	Key         string
+	KeyID       string
+}
+
+func getProfiles() (map[string]profile, error) {
+	rd, err := os.ReadFile(getCLIProfilesPath())
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		rd = make([]byte, 0)
+	}
+
+	profiles := make(map[string]profile)
+	if err := json.Unmarshal(rd, &profiles); err != nil {
+		return nil, err
+	}
+	return profiles, nil
+}
+
+func writeProfiles(profiles map[string]profile) error {
+	md, err := json.MarshalIndent(profiles, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	//nolint:gosec
+	return os.WriteFile(getCLIProfilesPath(), md, 0o640)
+}
+
+func addOrUpdateProfile(c *cli.Context, args addOrUpdateProfileArgs, isAdd bool) error {
+	profiles, err := getProfiles()
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		profiles = make(map[string]profile)
+	}
+
+	profile, alreadyExists := profiles[args.ProfileName]
+	if isAdd && alreadyExists {
+		return fmt.Errorf("attempted to add new profile %s but it already existed", args.ProfileName)
+	}
+	profile.APIKey.KeyCrypto = args.Key
+	profile.APIKey.KeyID = args.KeyID
+	profile.Name = args.ProfileName
+	profiles[args.ProfileName] = profile
+
+	if err := writeProfiles(profiles); err != nil {
+		return err
+	}
+
+	var addOrUpdate string
+	if isAdd {
+		addOrUpdate = "added"
+	} else {
+		addOrUpdate = "updated"
+	}
+
+	printf(c.App.Writer, "Successfully %s profile %s", addOrUpdate, args.ProfileName)
+	return nil
+}
+
+// AddProfileAction adds a new CLI profile.
+func AddProfileAction(c *cli.Context, args addOrUpdateProfileArgs) error {
+	return addOrUpdateProfile(c, args, true)
+}
+
+// UpdateProfileAction updates an existing CLI profile, or adds it if it doesn't already exist.
+func UpdateProfileAction(c *cli.Context, args addOrUpdateProfileArgs) error {
+	return addOrUpdateProfile(c, args, false)
+}
+
+type removeProfileArgs struct {
+	ProfileName string
+}
+
+func whichProfile(args *globalArgs) (_ *string, profileSpecified bool) {
```

Comment:
So I don't think you need named returns here. Also, is there a reason to return `*string` instead of just `string`?

Reply:
Yeah, I went with named returns because I worried that the significance of the return bool was ambiguous. I guess not too big a concern given the limited use here and the fact that it's not exported, I'll remove!

Re: `*string`, we check on nil-ness in `ConfigFromCache`. It could certainly be a regular string and we could check on if it's empty or not! I don't have a particular preference, do you think one is preferred?

Reply:
> I guess not too big a concern given the limited use here and the fact that it's not exported, I'll remove!

Yeah that comma-ok pattern you're using (`value, valueExists`) is common enough that I don't think you need the named returns.

> Re: *string, we check on nil-ness in ConfigFromCache. It could certainly be a regular string and we could check on if it's empty or not!

I had assumed you would check the bool return value (previously `profileSpecified`) and assume the `string` represented a profile only if that bool was true (as opposed to checking if `whichProf != nil`.) i.e. I think you have enough expressiveness of "existence" through the boolean return value such that the string only needs to be `string` and not `*string`. But, no super strong preference, and I may be misunderstanding what you're doing.

Reply:
Ahh, so the string might represent a profile even if `profileSpecified` is false, if the `VIAM_CLI_PROFILE_NAME` env var is set. Even when we know we have a profile, we care about whether it was specified via flag or not. specifically, we want to fallback on default behavior if the profile doesn't exist and _wasn't_ specified (maybe the env var was accidentally exported), but not if the profile doesn't exist and _was_ specified (since the user specifically asked for particular behavior, we shouldn't assume they'd be okay with different behavior).

Reply:
Gotcha, maybe a comment above `whichProfile` explaining exactly what you've said there could be helpful. Thanks for the extra context!